Update Citus Technical Documentation about the rebalancer (#7638)
citusdata/citus: src/backend/distributed/README.md

The shard rebalancer consists of 4 main parts:

1. The rebalancing algorithm: Decides what moves/splits it should do to make
   the cluster balanced.
2. The background task runner: Runs a full rebalance according to a plan
   created by the planner.
3. A shard group moves/split: These are the smallest units of work that the
   rebalancer does, if this fails midway through the move is aborted and the
   shard group remains unchanged.
4. Deferred cleanup: The source shards stay present for a while after a move to
   let long-running read queries continue, eventually they need to be cleaned
   up.

These parts interact, but they are pretty self-contained. Usually it's only
necessary to change one of them to add a feature/fix a bug.

[rebalancer-post]: https://www.citusdata.com/blog/2021/03/13/scaling-out-postgres-with-citus-open-source-shard-rebalancer/

## Rebalancing algorithm

The rebalancing algorithm tries to find an optimal placement of shard groups
across nodes. This is not an easy job, because this is a [co-NP-complete

> problem](https://en.wikipedia.org/wiki/Knapsack_problem). So instead going for

Reviewer: ```suggestion
problem](https://en.wikipedia.org/wiki/Knapsack_problem). So instead of going for
```

the fully optimal solution it uses a greedy approach to reach a local

> optimimum, which so far has proved effective in getting to a pretty optimal

Reviewer: ```suggestion
optimum, which so far has proved effective in getting to a pretty optimal
```

solution.

> Eventhough it won't result in the perfect balance, the  greedy aproach has two

Reviewer: ```suggestion
Even though it won't result in the perfect balance, the  greedy approach has two
```

1. It's relatively easy to understand why the algorithm decided on a certain move.
2. Every move makes the balance better. So if the rebalance is cancelled midway
   through, the cluster will always be in a better situation than before.

As described in the [this rebalancer blog post][rebalance-post] the algorithm
takes three inputs from the function in the `pg_dist_rebalance_strategy` table:

1. Is a shard group allowed on a certain node?
2. What is the "cost" of a shard group, relative to the other shard groups?
3. What is the "capacity" of a node, relative to the other nodes?

Cost and capacity are vague on purpose, this way users can choose their own
way to determine cost of a shard group, but **in practice "cost" is usually
disk size** (because `by_disk_size` is the default rebalance strategy).
Capacity is almost always set to 1, because almost all Citus clusters are
homogenous (they contain the same nodes, except for maybe the coordinator). The
main usage for "Is a shard group allowed on a certain node?" is to be able to pin a
specific shard group to a specific node.

There is one last definition that you should know to understand the algorithm

> and that is "utilization". Utilization is total cost of all shard groups

Reviewer: ```suggestion
and that is "utilization". Utilization is the total cost of all shard groups
```

divided by capacity. In practice this means that utilization is almost always
the same as cost because as explained above capacity is almost always 1. So if
you see "utilization" in the algorithm, for all intents and purposes you can
read it as "cost".

The way the general algorithm works is fairly straightforward. It starts by
creating an in-memory representation of the cluster, and then it tries to
improve that in-memory representation by making theoretical moves. So to be

> clear the algorithm doesn't actually do andy shard group moves, it only does

Reviewer 2: ```suggestion
clear the algorithm doesn't actually do any shard group moves, it only does
```

those moves to its in-memory representation. The way it determines what
theoretical moves to make is as follows (updating utilization of in-memory
nodes after every move):

1. Find all shard groups that are on a node where they are not allowed (due to
   "Is a shard group allowed on a certain node?")
2. Order those nodes by cost
3. Move them one-by one to nodes with the lowest utilization where they are
   allowed.
4. If the cluster is balanced we are done.
5. Take the most utilized node (A) and take the least utilized node (B).
6. Try moving the shard group with the highest cost from A to B.
7. If the balance is "better" commit this move and continue from step 4. (See subsection below for
   what is "better")
8. If the balance is worse/equal try again from step 6 with the shard group
   with the next highest cost on node A. If this was the lowest cost shard on
   node A, then try with the highest cost shard again but on the next least
   utilized node after node B. If no moves helped with the balance, try with
   the next most utilized node after node A. If we tried all moves for all
   nodes like this, we are done (we cannot get a better balance).

> Ofcourse the devil is in the details though.

Reviewer 2: ```suggestion
Of course, the devil is in the details though.
```

### When is the balance better?

The main way to determine if the balance is better is by comparing the
utilization of node A and B, before and ofter the move and seeing if they are
net closer to the average utilization of the nodes in the cluster. The easiest
way to explain this is with a simple example:

We have two nodes A and B. A has a utilization of 100GB and B has a utilization
of 70GB. So we will move a shard from A to B. A move of 15GB is obviously best,
it results in perfect balance (A=85GB, B=85GB). A move of a 10GB is still
great, both improved in balance (A=90GB, B=80GB). A move of 20GB is also good,
the result is the same as a move of 10GB only with the nodes swapped (A=80GB,
B=90GB).

The 10GB vs 20GB move shows a limitation of the current algorithm. The
algorithm mostly makes choices based on the end state, not on the cost of
moving a shard. This is usually not a huge problem in practice though.

### Thresholds

> The algorithm is full of thresholds, the main reason these exists is because

Reviewer 2: ```suggestion
The algorithm is full of thresholds, the main reason these exist is because
```

moving shards around isn't free.

- `threshold`: Used to determine if the cluster is in a good enough state. For
  the `by_disk_size` rebalance strategy this is 10%, so if all nodes are at
  most 10% above or 10% below the average utilization then no moves are
  necessary anymore (i.e. the nodes are balanced enough). The main reason for
  this threshold is that these small differences in utilization are not

>   necessarily problematic and might very well resolve automatically over time

Reviewer 2: ```suggestion
  necessarily problematic and might very well resolve automatically over time. For example, consider a scenario in which
```

`rebalance_table_shards()`, this function run the rebalance using the current
session. This has the huge downside that the connection needs to be kept open
until the rebalance completes. So eventually we [introduced
`citus_rebalance_start()`](https://www.citusdata.com/blog/2022/09/19/citus-11-1-shards-postgres-tables-without-interruption/#rebalance-background),
which uses a background worker to do the rebalancing, so users can disconnect
their client and the rebalance continues. It even automatically retries moves
if they failed for some reason.

The way this works is using a general background job infrastructure that Citus
has in the tables `pg_dist_backround_job` and `pg_dist_background_task`.
A job (often) contains multiple tasks. In case of the rebalancer, the job is
the full rebalance, and each of its tasks are separate shard group moves.

### Parallel background task execution

A big benefit of the bacground task infrastructure is that it can execute tasks
and jobs in parallel. This can make rebalancing go much faster especially in
clusters with many nodes. To ensure that we're not doing too many tasks in
parallel though we have a few ways to limit concurrency:

> 1. Tasks can depend on eachother. This makes sure that one task doesn't start

Reviewer: ```suggestion
1. Tasks can depend on each other. This makes sure that one task doesn't start
```
